### ISSUE TITLE
`setarch` in Linux launcher breaks in podman containers

### ISSUE THREAD

**Issue by draeath** [DFHack/dfhack#1466]
dfhack is unable to be executed in a container due to the usage of `setarch`. The following is returned:

```
setarch: failed to set personality to x86_64: Operation not permitted
```

Is there a particular reason this is being done?

**Comment by draeath** [comment]
For some context, this came up while experimenting with running DF within a rootless podman container.

**Comment by draeath** [comment]
Ah, I see. This is being done to disable address space randomization and you can't seem to run setarch at all without specifying a personality.

In this case, `setarch linux64 -R` works, where `x86_64` fails. I can't think of any reason not to use linux64 over x64_64. That might be a suitable fix?

**Comment by lethosor** [comment]
We don't just put random commands in the launcher script for fun. It looks like `setarch` was introduced in https://github.com/DFHack/dfhack/commit/c423bb6e8858559d044f21fb412dabe09ffbe35c for the purposes of disabling ASLR, which is extremely important, since DFHack on non-Windows can't handle ASLR.

**Comment by draeath** [comment]
Please excuse my tone, I wasn't intending to imply it was done for giggles.

Also note I added a comment with just that realization about the same time you replied.

Using 'linux64' instead of 'x86_64' looks to solve the issue in containers, but I am not 100% if this will have any detrimental effect outside of one.

**Comment by lethosor** [comment]
Yeah, I just saw your reply, thanks.

> In this case, `setarch linux64 -R` works, where `x86_64` fails. I can't think of any reason not to use linux64 over x64_64. That might be a suitable fix?

Is there a list of valid architectures somewhere? This is the first time I've seen `linux64` mentioned as a valid option, and I'm wary of changing to that in case other systems don't support that. Plus, the DF executable is specifically built for x86-64 - is your container able to run DF?

The argument to `setarch` is currently determined at build time, because reliably detecting the architecture at runtime across distros/etc. proved to be difficult. `setarch --list` could work, though, assuming that's portable.

**Comment by draeath** [comment]
... and now it has stopped working, and I don't understand what changed in my environment to do so.

But it *did* work for several starts. Not sure what the heck is going on, there, or how to properly mitigate it beyond yelling at the Moby or CRI-O projects to allow "changing" the personality.

**Comment by draeath** [comment]
As far as portability of `setarch --list` - this was added to the upstream util-linux codebase [about 5 years back.](https://github.com/karelzak/util-linux/commit/5edb0ea6bbd57dd916417737f98c9109dc1ecb5b)

**Comment by lethosor** [comment]
I guess another thing we could try is `setarch some_arch -R true` to find an architecture that works before starting DF. Checking this on a Pi is an interesting idea too. I'll leave this open until we can come up with something a bit more robust.

In your case, I expect you can change the contents of `hack/dfhack_setarch.txt` to an architecture that works in your container (if you're able to find one... if it only works intermittently, I'm not sure how to deal with that from our end)

**Comment by draeath** [comment]
Someone has covered this ground on the container runtime side and there appears to be a solution that documented [here](https://github.com/BenLubar/df-docker/blob/3c08fafbadfd60788e12d6a9e0e11c05f4ed751b/README.md).

Again though, this is out of scope for dfhack I think, unless we can find some *other* way to disable ASLR for a binary that doesn't depend on setarch or kernel tunables.

Thanks for your time!

**Comment by orynider** [comment]
How a Windows version should be?

**Comment by myk002** [comment]
Windows builds of DFHack don't use the Linux launcher
